### PR TITLE
bpo-41682: fixed flaky test test_sendfile_close_peer_in_the_middle_of_receiving

### PR DESCRIPTION
**Pull request by kumaraditya303**
https://bugs.python.org/issue41682

### CODE REVIEW

**Comment by kumaraditya303** [comment]
...And we finally have green CI 🎉

**Comment by kumaraditya303** [comment]
cc @gvanrossum @vstinner @pablogsal

**Review by erlend-aasland — APPROVED** [review]
Assuming Guido's reasoning is correct, this is IMO the correct fix. LGTM!

(But note that this test has been flakey for two years; maybe there was another reason for the failures before?)

**Comment by erlend-aasland** [comment]
> ...And we finally have green CI 🎉

Actually, it has been green since Nikita's #30801 😉 Please sync with `main` and remove the Windows guard added by #30801. If this fixes the test on Windows, we no longer need the guard.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]
> ...And we finally have green CI 🎉

Actually, it has been green since Nikita's #30801 😉 Please sync with `main` and remove the Windows guard added by #30801. If this PR fixes the test on Windows, we no longer need the guard.

**Comment by kumaraditya303** [comment]
> Assuming Guido's reasoning is correct, this is IMO the correct fix. LGTM!
> 
> (But note that this test has been flakey for two years; maybe there was another reason for the failures before?)

I was already trying to find the size of DATA based on some stack overflow answers before Guido provided his reasoning as I once tried making DATA very high and it worked so I knew it was about fine tuning the DATA length.

**Comment by erlend-aasland** [comment]
> I already removed the skip for windows, no?

Great, I see it. Thanks.

**Review by erlend-aasland — APPROVED** [review]

File: Lib/test/test_asyncio/test_sendfile.py [hunk 1]
```diff
@@ -92,9 +92,13 @@ async def wait_closed(self):
 
 class SendfileBase:
 
-      # 128 KiB plus small unaligned to buffer chunk
-    DATA = b"SendfileBaseData" * (1024 * 8 + 1)
-
+    # 256 KiB plus small unaligned to buffer chunk
+    # Newer versions of Windows seems to have increased its internal 
+    # buffer and tries to send as much of the data as it can as it
```
Comment (erlend-aasland) [inline Lib/test/test_asyncio/test_sendfile.py:97]: Nit: I'd try to verify this claim. Perhaps by adding a link to MS API docs, or something similar.

**Comment by Fidget-Spinner** [comment]
I'm gonna test with buildbots. For now, please hold off on committing anything (well you can, but it'll make the buildbots status harder to track ;).

If this works, we need to give you an award for saving thousands of contributors from this pain.

**Comment by erlend-aasland** [comment]
> If this works, we need to give you an award for saving thousands of contributors from this pain.

We could consider giving this patch a NEWS entry, given the amount of frustration and gray hair this test has caused 😆

**Comment by gvanrossum** [comment]
I'll leave this in Erlend's capable hands.

**Review by asvetlov — APPROVED** [review]

File: Lib/test/test_asyncio/test_sendfile.py (hunk 1 above)
Comment (asvetlov) [reply Lib/test/test_asyncio/test_sendfile.py:97]: Honestly, I doubt if it is reflected in docs. 
Moreover, the socket buffer size seems to depend on the OS build version.

**Comment by erlend-aasland** [comment]
Thanks for you input, @asvetlov!

I would make the PR title and message more accurate before merging; maintaining a clear and self-documenting git history really helps when issues pop up again (I suspect this might do in a Windows release or five). Here's a quick 'n dirty suggestion, but feel free to word it differently:

```
bpo-41682: Double the size of test_asyncio.test_sendfile data buffer

Some Windows versions use a large internal transmit buffer, which can
make tests that rely on fragmented transmits flakey (for example
test_sendfile_close_peer_in_the_middle_of_receiving).

We now use a buffer slightly larger than 256 KiB.
```

Other than that, I have no further remarks. (The failed buildbot is not related to this PR.)

Here's your CI Janitor Award: 🏆🥇 Thanks for keeping the bots green! Thanks, Guido and Kumar, for the reasoning and the patch.

(Andrew, Guido, or Ken can merge; I don't have the commit bit.)

**Comment by erlend-aasland** [comment]
(The bot that edit our PR comments in order to create bpo links can be _very, very_ irritating at times...)

**Comment by pablogsal** [comment]
Merged ;)

**Comment by vstinner** [comment]
> @pablogsal: Merged ;)

Be careful next time you merge a PR, the merged commit title is just "fixed flaky test" which doesn't mention test_asyncio or the bpo number, so no comment was added to https://bugs.python.org/issue41682

That's the annoying part with GH PR: it's possible to change the title and description in GitHub, and then get I completly different title and description when you merge it...

**Comment by pablogsal** [comment]
> > @pablogsal: Merged ;)
> 
> Be careful next time you merge a PR, the merged commit title is just "fixed flaky test" which doesn't mention test_asyncio or the bpo number, so no comment was added to https://bugs.python.org/issue41682
> 
> That's the annoying part with GH PR: it's possible to change the title and description in GitHub, and then get I completly different title and description when you merge it...

I also used the GitHub phone app to merge it which o think made things even more confusing as I didn't see that the commit title would be different :(

**Comment by vstinner** [comment]
I agree, the UI on a phone is "not great" :-( Especially for a merge operation.

**Pull request merged**